timbermill-client/src/main/java/com/datorama/timbermill/pipe/TimbermillServerOutputPipe.java

```diff
@@ -0,0 +1,170 @@
+        maxSecondsBeforeBatchTimeout = builder.maxSecondsBeforeBatchTimeout;
+        buffer = Collections.synchronizedList(new LinkedList<>());
+    }
+
+    private void startThread() {
+        thread = new Thread(() -> {
+            do {
+                try {
+                    List<Event> eventsToSend = createEventsToSend();
+                    if (eventsToSend != null) {
+                        EventsWrapper eventsWrapper = new EventsWrapper(eventsToSend);
+                        HttpURLConnection httpCon = getHttpURLConnection();
+                        byte[] eventsWrapperBytes = getEventsWrapperBytes(eventsWrapper);
+                        sendEventsOverConnection(httpCon, eventsWrapperBytes);
+                        int responseCode = httpCon.getResponseCode();
+                        if (responseCode != 200){
+                            LOG.warn("Request to timbermill return status {}, Message: {}", responseCode, httpCon.getResponseMessage());
+                        }
+                        else{
+                            LOG.debug("{} were sent to Timbermill server", eventsToSend.size());
+                        }
+                    }
+                } catch (Exception e) {
+                    LOG.error("Error sending events to Timbermill server" ,e);
+                }
+            } while (buffer.size() > 0);
+        });
+        thread.start();
+    }
+
+    private void sendEventsOverConnection(HttpURLConnection httpCon, byte[] eventsWrapperBytes) throws IOException {
+        OutputStream os = httpCon.getOutputStream();
+        os.write(eventsWrapperBytes);
+        os.flush();
+        os.close();
+    }
+
+    private byte[] getEventsWrapperBytes(EventsWrapper eventsWrapper) throws JsonProcessingException {
+        ObjectMapper om = new ObjectMapper();
+        ObjectWriter ow = om.writerFor(om.getTypeFactory().constructType(EventsWrapper.class));
+        String eventsWrapperString = ow.writeValueAsString(eventsWrapper);
+        return eventsWrapperString.getBytes();
+    }
+
+    private HttpURLConnection getHttpURLConnection() throws IOException {
+        HttpURLConnection httpCon = (HttpURLConnection) timbermillServerUrl.openConnection();
+        httpCon.setRequestMethod("POST");
+        httpCon.setRequestProperty("content-type", "application/json");
+        httpCon.setDoOutput(true);
+        return httpCon;
+    }
+
+    private List<Event> createEventsToSend() throws JsonProcessingException {
+        long startBatchTime = System.currentTimeMillis();
+        List<Event> eventsToSend = new ArrayList<>();
+        int currentBatchSize = 0;
+        if(!buffer.isEmpty()) {
+            Event event = buffer.remove(0);
+            currentBatchSize += getEventSize(event);
+            eventsToSend.add(event);
+
+            while(!buffer.isEmpty() && currentBatchSize <= this.maxEventsBatchSize && !isExceededMaxTimeToWait(startBatchTime)) {
```

Comment:
consider dropping the isExceededMaxTimeToWait check, not sure it's needed

Reply:
It's needed so we won't keep events in memory forever without sending them. If we reach this timeout, stop and send what you have.